src/compiler.ts

```diff
@@ -3236,6 +3243,7 @@ export class Compiler extends DiagnosticEmitter {
     // Compile the body assuming the condition turned out true
     var bodyFlow = flow.fork();
     bodyFlow.inheritNonnullIfTrue(condExpr);
+    bodyFlow.inheritNarrowedTypeIfTrue(condExpr);
```

Comment:
It seems to me that once narrowing is in place, that the nonnull machinery becomes obsolete. Hence wondering if it would be sufficient to upgrade the nonnull machinery to narrowing so there is only one concept in the end?

Reply:
Yes, and it can resolve the nonull bug mentioned in the last of #2360.
But I think maybe we can do it one by one instead of doing everything in this PR? It maybe make this PR very hard to review